x-pack/plugins/actions/server/builtin_action_types/lib/send_email.ts

```diff
@@ -75,25 +81,58 @@ export async function sendEmail(logger: Logger, options: SendEmailOptions): Prom
 async function sendEmailWithExchange(
   logger: Logger,
   options: SendEmailOptions,
-  messageHTML: string
+  messageHTML: string,
+  connectorTokenClient: ConnectorTokenClient
 ): Promise<unknown> {
-  const { transport, configurationUtilities } = options;
+  const { transport, configurationUtilities, connectorId } = options;
   const { clientId, clientSecret, tenantId, oauthTokenUrl } = transport;
-  // request access token for microsoft exchange online server with Graph API scope
 
-  const tokenResult = await requestOAuthClientCredentialsToken(
-    oauthTokenUrl ?? `${EXCHANGE_ONLINE_SERVER_HOST}/${tenantId}/oauth2/v2.0/token`,
-    logger,
-    {
-      scope: GRAPH_API_OAUTH_SCOPE,
-      clientId,
-      clientSecret,
-    },
-    configurationUtilities
-  );
+  let accessToken: string;
+
+  const connectorToken = await connectorTokenClient.get({ connectorId });
+  if (connectorToken === null || Date.parse(connectorToken.expiresAt) <= Date.now()) {
+    // request new access token for microsoft exchange online server with Graph API scope
+    const tokenResult = await requestOAuthClientCredentialsToken(
+      oauthTokenUrl ?? `${EXCHANGE_ONLINE_SERVER_HOST}/${tenantId}/oauth2/v2.0/token`,
+      logger,
+      {
+        scope: GRAPH_API_OAUTH_SCOPE,
+        clientId,
+        clientSecret,
+      },
+      configurationUtilities
+    );
+    accessToken = `${tokenResult.tokenType} ${tokenResult.accessToken}`;
+
+    // try to update connector_token SO
+    try {
+      if (connectorToken === null) {
+        await connectorTokenClient.create({
+          connectorId,
+          token: accessToken,
+          expiresAt: new Date(Date.now() + tokenResult.expiresIn).toISOString(),
```

Comment:
I think this is off too. `Date.now()` returns `ms` but the `expiresIn` response is [in seconds](https://docs.microsoft.com/en-us/graph/auth-v2-service#token-response)

Reply:
Awesome catch! Thank you.